### ISSUE TITLE
Simplify datastructure

### ISSUE THREAD

**Issue by LecrisUT** [spglib/spglib#489]
For the high-level interface Python, C++, (maybe even Fortran), we should discuss having a more structured data type for the input and output. I have touched on this in #301 briefly, but we should finish this at some point to move forward. We can start providing this interface first in Python since it is the more popular. For this interface we need to distinguish the base data from the interface or other indirect data.

Let's keep the top-most comment up-to-date with the most recent ideas

### Input: Lattice + Atom

Basic input structure
```python
@dataclasses.dataclass
class Lattice:
    lattice: numpy.array[float]  # shape(3,3)
    atoms: list[Atom]

@dataclasses.dataclass
class Atom:
    # Can this class be made into numpy type?
    position: numpy.array[float]  # shape(3,)
    type: int
```

### Output: Symmetry + SpaceGroupType

These are the basic output elements

```python
@dataclasses.dataclass
class Symmetry:
    rotation: numpy.array[float]  # shape(3,3)
    translation: numpy.array[float]  # shape(3,)

@dataclasses.dataclass
class SpaceGroupType:
    number: int
```

#### Interfaces

```python
class Symmetry:
    def is_rotation(self) -> bool:
        pass
    def is_translation(self) -> bool:
        pass

class SpaceGroupType:
    @property
    def international_symbol(self) -> str:
        pass
    @property
    def hall_symbol(self) -> str:
        pass
    @property
    def pointgroup_international(self) -> str:
        pass
    @property
    def pointgroup_schoenflies(self) -> str:
        pass
    @property
    def arithmetic_crystal_class_symbol(self) -> str:
        pass
```

Helper classes like `RotationSymmetry`, `TranslationSymmetry` can be added to more easy filtering

### Output: Dataset

Top-most output interface that contains most interfaces
```python
@dataclasses.dataclass
class Dataset:
    symmetries: list[Symmetry]
    spacegroup: SpaceGroupType
    lattice: Lattice
```

#### Interfaces

```python
class Dataset:
    @property
    def primitive_lattice(self) -> Lattice:
        pass
    @property
    def transformation_matrix(self) -> numpy.array[float]:  # shape(3,3)
        pass
    @property
    def origin_shift(self) -> numpy.array[float]:  # shape(3,)
        pass
```

### Magnetic equivalents

TODO

## To discuss

- Namings
- Help messages
- Interfaces
- Minimal data structure
- What about mixed rotation + translation symmetry?
- Deprecated interfaces

---

PS: This is not part of the 2.5.0 introduction of `dataclass`, this will come after that

**Comment by LecrisUT** [comment]
@atztogo @lan496 Could these interface be refined to align with `moyo`'s vision?

**Comment by atztogo** [comment]
@LecrisUT, do spglib have to align with moyo?

**Comment by LecrisUT** [comment]
If spglib alligns with moyo it would help give users a few release time to migrate. The inverse of moyo carrying compatibility layers to spglib is also possible, but it would add maintenance burden on moyo, push migration window later, and miss out on spglib user feedback of the interface they desire.

**Comment by atztogo** [comment]
This sits at the boundary between science and software engineering, which makes it hard to judge, but I think it's better to wait until a significant problem actually arises in practice.

**Comment by LecrisUT** [comment]
> but I think it's better to wait until a significant problem actually arises in practice.

The issue that this is trying to avoid is making a hard transition for the users. I have seen this issue play out a lot where a 20 year old code is still being packaged in Fedora because one dependent package could not update to the author's recommended replacement. And in academia this issue is a magnitude worse.

> This sits at the boundary between science and software engineering, which makes it hard to judge

Not really there is already a moyopy interface that is being worked on. The issue here is to lock in what works well there and highlight those. I recommend the early discussion while the code is still flexible to avoid alienating the users.

**Comment by atztogo** [comment]
You have a fair point, but we also need to pace ourselves so we don't burn out. If that happens, the project won't make it to the finish line, and that's what worries me most.

**Comment by lan496** [comment]
I agree that the high-level data model is worth discussing, but I do not yet see a clear motivation for introducing a new input dataclass on the spglib side.

On the moyopy side, I have already started writing a migration guide from spglib: https://github.com/spglib/moyo/blob/main/moyopy/docs/migration.md . At least for the current Python interface, we can still keep an almost one-to-one correspondence with spglib.  `moyo.Cell` can be understood as essentially the same kind of object as the current spglib cell: basis vectors, fractional positions, and atom types. Crystallographically, this is close to the definition of a crystal pattern: https://dictionary.iucr.org/Crystal_pattern . 

If we split the data model further, I think the more natural separation would be:
1. domains: basis vectors + positions
2. atom types

rather than introducing a `Lattice` dataclass containing a list of `Atom`s. So my current preference is to keep the input representation close to the existing spglib cell unless we identify a concrete limitation of the current model.

**Comment by LecrisUT** [comment]
First let me clarify the expectations here:
- The purpose of this issue is not to state a specific interface to adopt, but to open a discussion on potential improvements to the interface both internally and at the API level
- Adopting one datastructure would not mean that *only* that structure is supported and mirrored internally. Translating between one and another structure can be very cheap, particularly at the high interface level
- A big proponent to this proposal is to allow for more inheritance between the common data and to allow users to build up on the basic structure, or become compatible with their internal structure

---

With that said, let me break down the reasoning behind the structure initially proposed here:

#### Grouping of the Symmetry

This is a major UI improvement. Currently the `rotations`, `translations` and `time_reversals` are distinct `Nx3x3`, `Nx3`, `N` tensors. This is very clunky for the user as they have to navigate all of these tensors independently to find the specific symmetry operation `(rotation, translation, time_reversal)` to extract. But that's the thing, these are a whole group of data that is better represented as `N` symmetry operations each with a `3x3`, `3`, `1` tensors.

#### Data locality

The information of `position`, `numbers`, `magnetic_moments` are always tied together. Having these 3 grouped together helps with the cpu cache. Consider the following snippet
https://github.com/spglib/spglib/blob/12355c77fb7c505a55f52cae36341d73b781a065/src/symmetry.c#L537-L546
In the worse case of the current model, `position[i_atom]`, `position[j]`, `types[i_atom]`, `types[j]`, all 4 can be distant in memory space due to the lack of correlation between `position` and `types`. Contrast that with `atoms[i_atom].position`, `atoms[i_atom].type` are guaranteed to be local to each (with appropriate internal datastructure types), and at most you would have only 2 distant memory spaces with `atoms[j].position`/`atoms[j].type`.

Extracting a vector of `position`, `numbers`, `magnetic_moments` is possible from a vector of `atoms` without reshuffling and copying data, but not so the other way around.

#### Cleaner interface and less data copies

Consider the interface in the function being called above
https://github.com/spglib/spglib/blob/12355c77fb7c505a55f52cae36341d73b781a065/src/cell.c#L203-L206

And this principle applies across compiled languages C, C++, Fortran, and Rust (afaiu that language). Python a bit less so since it *always* uses references.

#### Inheritance within the internal structure

There is an inherent subclass relationship between the spatial symmetry operations and the magnetic one, namely
```python
@dataclass
class Atom:
    position: list[float]
    type: int

@dataclass
class MagneticAtom(Atom):
    magnetic_moment: float

@dataclass
class Symmetry:
    rotation: list[list[float]]
    translation: list[float]

@dataclass
class MagneticSymmetry(Symmetry):
    time_reversal: list[bool]
```
This will help deduplicate internal code as OOP can take a covariant vector of a base class `Atom` and share the function internally.

#### User extensions

The user can have various other data attached to an atom like
```python
@dataclass
class UserAtom(Atom):
    symbol: str
    atomic_number: int
```
Having the user translate to-and-from spglib data is non-trivial. On the other hand, this data structure can be preserved via a `copy` operation, changing only the internal data, etc.

What's more the user does not have to use our specific dataclass class. We just need to define a `Protocol` of the expected data, and use the `getattr` to navigate the data.

#### Minimizing the operations

Consider parts like `std_lattice`, does the user always need it, or can it be calculated on demand. This may not be the best example, but the principle is there, what is the bare minimum data required to work with and what are indirect properties, or function interfaces.

#### Dropping unnecessary baggage

This is an opportunity to consider all of the baggage that has piled up in `spglib` and is being carried over to moyo. Not all of the properties make sense to remain, and can be deprecated out to clean out the interface. I only needed to take a quick look at the moyo interface to find one, `num_atoms`
https://github.com/spglib/moyo/blob/f6be9b22699691caa87379c16e744bae06c39069/moyopy/python/moyopy/_base.pyi#L31-L33
This is not a useful user information and instead can easily be extracted with a `len()` operation.

---

Does all of the above mean that I am suggesting to scrap everything and go back to the drawing board? No! The goal for the above is to consider what are the implications of the current API design and consider what alternatives would help both internally and the users. The only code change that would be associated with this is to experiment with exposing a new user interface that would map it to whatever the internal structure is. Anything further would just come as we and the users see potential in these.

**Comment by lan496** [comment]
This is not central to the main point, but I am not sure cache locality should be the main consideration when discussing SoA vs AoS here. I do not expect spglib to become a bottleneck in that sense. That said, I agree with the general direction of reducing raw indexing by adding more methods to the current `Cell`-like structure.

For `Symmetry`, I think the situation is a bit subtle. Representing each symmetry operation as a class containing `(rotation, translation, time_reversal)` is certainly more OOP-like. However, in practice, there are also cases where users want direct access to a list/array of rotations, for example when extracting only the point group. So I am not sure that making the data structure more “proper” necessarily improves usability in all cases.

I am not sure whether this can be realized cleanly within the current simple C-style structure, but providing an iterator over `(rotation, translation, time_reversal)` sounds like a good compromise to me.

**Comment by LecrisUT** [comment]
> However, in practice, there are also cases where users want direct access to a list/array of rotations, for example when extracting only the point group. So I am not sure that making the data structure more “proper” necessarily improves usability in all cases.

Of course, perfectly valid. I do not propose to eliminate one in favor of another, these interfaces can live alongside each other and the user chooses whichever is more convenient for them at a specific moment. This can live as high-level interface even if it would be as wrappers.

> I am not sure whether this can be realized cleanly within the current simple C-style structure, but providing an iterator over `(rotation, translation, time_reversal)` sounds like a good compromise to me.

Definitely not at the C level or the bindings. Only at the user-facing OOP frontends. About the tuple suggestion, I wish to avoid those as much as possible, they are unnavigable, undocumented, and inflexible. A tuple can be extracted from a class, but not the other way around. That said, yes, an iterator of the structured data in the current interface as a combination of the 3 data is also a good example of a simple way forward  to experiment with this idea
